Prefer paths with more liquidity when qualities are equal

seelabs opened the pull request:
This PR has three changes:

1) When evaluating path qualities in flow V2, prefer paths with more liquidity when qualities are equal.

2) Added debugging probes that help me compare flow v1 and flow v2 differences. I intend to take out these debugging probes when we remove flow v1. I also added a feature to enable these comparisons, so they are off by default

3) I took flow V2 off a date switch, it will be enabled with an amendment only

scottschurr commented on src/ripple/app/paths/impl/StrandFlow.h:
```diff
@@ -468,7 +468,8 @@ flow (PaymentSandbox const& baseView,
 
-            if (!best || q > best->quality)
+            if (!best || best->quality < q ||
+                (best->quality == q && best->out < f.out))
```
Nice that the change only impacted one line.

scottschurr commented on src/ripple/ledger/PaymentSandbox.h:
```diff
@@ -188,6 +188,13 @@ class PaymentSandbox final
     apply (PaymentSandbox& to);
     /** @} */
 
+    // Return a map of balance changes on trust lines and amount of XRP destoryed.
```
"destoryed" -> "destroyed"?

scottschurr commented on src/ripple/ledger/PaymentSandbox.h:
```diff
@@ -188,6 +188,13 @@ class PaymentSandbox final
     /** @} */
 
+    // Return a map of balance changes on trust lines and amount of XRP destoryed.
+    // The low account is be the first account in the key.
```
"is be" -> "is"?

scottschurr commented on src/ripple/ledger/PaymentSandbox.h:
```diff
@@ -188,6 +188,13 @@ class PaymentSandbox final
+    // Return a map of balance changes on trust lines and amount of XRP destoryed.
+    // The low account is be the first account in the key.
+    // If the two accounts are equal, the map contains the total changes in currency
+    // regargless of issuer. This is useful to get the total change in XRP balances.
```
"regargless" ->? "regardless"?

seelabs replied:
I will spell check all the strings and comments in this file, thanks for find these.

scottschurr commented on src/ripple/ledger/PaymentSandbox.h:
```diff
@@ -188,6 +188,13 @@ class PaymentSandbox final
+    // The low account is be the first account in the key.
+    // If the two accounts are equal, the map contains the total changes in currency
+    // regargless of issuer. This is useful to get the total change in XRP balances.
+    std::pair<std::map<std::tuple<AccountID, AccountID, Currency>, STAmount>, XRPAmount>
```
Not excited about returning a pair here.  To me it makes more sense to have two different methods: `balancesChanges` returns the map and a `dropsDestroyed` method to return the destroyed XRP.  The implementation doesn't suggest that there's an optimization from computing both together (although I could be wrong on that).  "5. Give one entity one cohesive responsibility" -- Sutter and Alexandrescu

scottschurr replied:
@vinniefalco: Yeah, these signatures will be ugly.  But if we can sensibly disentangle a bit of it that will help.  If there's a performance benefit to the `pair` return value then it makes sense to leave this as it is.

seelabs replied:
@scottschurr @vinniefalco This function is used to find metadata differences between the two versions of payments. This data is only collected when a particular feature is enabled, and put into a `boost::optional`. My idea was if I collected additional metadata (offers changed, say), then I would add it to this one function.

I do think @scottschurr has a point about separating out the two methods, and will do so. However, when they are called they will likely be called from a lambda that returns a pair, so that the caller keeps a single `optional` around for all the collected metadata.

vinniefalco replied:
I could go either way about it

scottschurr replied:
@seelabs: I don't feel strongly about it and you have a better view of the larger context.  Do what you feel is appropriate.

scottschurr commented on src/ripple/ledger/PaymentSandbox.h:
```diff
@@ -188,6 +188,13 @@ class PaymentSandbox final
+    // If the two accounts are equal, the map contains the total changes in currency
+    // regargless of issuer. This is useful to get the total change in XRP balances.
+    std::pair<std::map<std::tuple<AccountID, AccountID, Currency>, STAmount>, XRPAmount>
+    balancesChanges (ReadView const& view) const;
```
`balancesChanges` may have been an intentional name, but perhaps `balancesChanged` would read a bit more naturally?  Just a thought.

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -260,4 +262,138 @@ PaymentSandbox::apply (PaymentSandbox& to)
+
+    // populate a dictionary with low/high/currency/delta // this can be
+    // compared with the old flow, and xrp deltas// can be checked with amount
+    // destroyed
```
The "//" occurrences in this paragraph look like there may be a comment wrapping peculiarity.  Are they intentional?

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -260,4 +262,138 @@ PaymentSandbox::apply (PaymentSandbox& to)
+                    highID = (*before)[sfAccount];
+                    break;
+                case ltRIPPLE_STATE:
+                    {
```
I don't think the braces here are necessary, since we're only assigning to pre-existing locals.

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -260,4 +262,138 @@ PaymentSandbox::apply (PaymentSandbox& to)
+                        lowID = (*before)[sfLowLimit].getIssuer();
+                        highID = (*before)[sfHighLimit].getIssuer();
+                        oldBalance = (*before)[sfBalance];
+                        highID = (*before)[sfAccount];
```
Um, hold it.  There are two assignments to `highID` here and no assignments to `newBalance`.  Is that intentional?

seelabs replied:
That looks like a copy/paste error. We're deleting here, so `newBalance` is zero, it should be `oldBalance.zeroed()`; As-is  it's a default constructed `STAmount` (which does construct to zero, so it worked, but I just got lucky). Good eyes.

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -260,4 +262,138 @@ PaymentSandbox::apply (PaymentSandbox& to)
+                    highID = (*after)[sfAccount];
+                    break;
+                case ltRIPPLE_STATE:
+                    {
```
Ultra minor: I think these braces could be removed.

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -260,4 +262,138 @@ PaymentSandbox::apply (PaymentSandbox& to)
+        auto delta = newBalance - oldBalance;
+        auto const cur = newBalance.getCurrency();
+        result[std::make_tuple(lowID, highID, cur)] = delta;
+        auto r = result.emplace(std::make_tuple(lowID, lowID, cur), delta);
```
Does the
```result[std::make_tuple(lowID, highID, cur)] = delta;```
duplicate the effort of the
```auto r = result.emplace(std::make_tuple(lowID, lowID, cur), delta);```?
I wonder if you could remove the line that array indexes into `result`.

seelabs replied:
@scottschurr I don't think so. Notice those are different keys. One is `low,high,cur`, the other is `low, low, cur`, the other is `high, high,cur`.

scottschurr replied:
@seelabs:  Your eyes are better than mine.  Thanks for pointing that out.

scottschurr commented:
On an OS X clang.debug.nounity build I'm getting a duplicate symbol error at link time:
```
duplicate symbol __ZN6ripple4path6detail12balanceDiffsERKNS_14PaymentSandboxERKNS_8ReadViewE in:
    build/clang.debug.nounity/src/ripple/app/paths/Flow.o
    build/clang.debug.nounity/src/ripple/app/paths/RippleCalc.o
ld: 1 duplicate symbol for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I can investigate after I finish the code review.

seelabs commented:
@scottschurr The non-unity link error is likely because I forgot to put make the new `balanceDiffs` function inline. I'll confirm here then push.

nbougalis commented on src/ripple/app/paths/RippleCalc.cpp:
```diff
@@ -129,48 +137,79 @@ RippleCalc::Output RippleCalc::rippleCalculate (
+            bool outputPassInfo,
+            bool outputBalanceDiffs) {
+            j.debug () << "RippleCalc Result> " <<
+            " actualIn: " << result.actualAmountIn <<
```
Please indent this one level.

scottschurr commented on src/ripple/app/paths/impl/FlowDebugInfo.h:
```diff
@@ -0,0 +1,364 @@
+        {
+        }
+        bool nativeIn;
+        bool nativeOut;
```
Could `nativeIn` and `nativeOut` be `const`?

scottschurr commented on src/ripple/app/paths/impl/FlowDebugInfo.h:
```diff
@@ -0,0 +1,364 @@
+        void
+        pushLiquiditySrc (EitherAmount const& in, EitherAmount const& out)
+        {
+            assert(!liquiditySrcIn.empty());
```
I'm confused.  Why does this `assert` not fire the first time `pushLiquiiditySrc` is called?  Thanks.

scottschurr replied:
Never mind.  It's the call to `resize ` in `newLiquidityPass`.  Sorry I missed that.

seelabs replied:
@scottschurr Because it's a vector of vectors, one vector per pass. A new vector is pushed in `newLiquidityPass`. It is a logic error if `pushLiquiditySrc` is called before `newLiquidityPass`.

scottschurr commented on src/ripple/app/paths/impl/FlowDebugInfo.h:
```diff
@@ -0,0 +1,364 @@
+        return i->second;
+    }
+
+    // Time the duration of the existance of the result
```
"existence" -> "existence"?

seelabs replied:
Thanks, fixed in a previous commit (I ran this file through a spell check earlier after you found other errors).

nbougalis commented on src/ripple/app/paths/impl/StrandFlow.h:
```diff
@@ -468,7 +479,8 @@ flow (PaymentSandbox const& baseView,
+                (best->quality == q && best->out < f.out))
                 best.emplace (f.in, f.out, std::move (*f.sandbox), *strand, q);
         }
 
```
While we're in this code: the `bool const shouldBreak = !bool(best)` immediately below could simply become:

    if (best)

nbougalis commented:
I'm ok with this with the commit addressing @scottschurr's feedback.

scottschurr commented on src/ripple/ledger/View.h:
```diff
@@ -340,8 +340,8 @@ transferXRP (ApplyView& view,
 
-NetClock::time_point const& flowV2SoTime ();
-bool flowV2Switchover (NetClock::time_point const closeTime);
+NetClock::time_point const& numAccSoTime ();
```
`numAcc` is a little too abbreviated for me to make sense of.  Could there be a comment explaining what `numAcc` means?

vinniefalco replied:
The 1990s called, they want their variable names back.

seelabs replied:
naming a switchover that turns on a feature is easy. Naming switchovers that fix bugs is harder. In this case, the switchover improves numerical accuracy in some places, so I chose `numAccSwitchover`. I'll try to come up with a better name.

vinniefalco replied:
`amendmentRIPD1141`

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -294,19 +294,16 @@ PaymentSandbox::balancesChanges (ReadView const& view) const
+                    lowID = (*before)[sfLowLimit].getIssuer();
+                    highID = (*before)[sfHighLimit].getIssuer();
+                    oldBalance = (*before)[sfBalance];
+                    newBalance = oldBalance.zeroed();
```
Nice.  It's easier for me to think about when you modify the locals in the same order.  Good change IMO.

scottschurr commented on src/ripple/ledger/impl/PaymentSandbox.cpp:
```diff
@@ -266,10 +266,9 @@ std::map<std::tuple<AccountID, AccountID, Currency>, STAmount>
+    // get the change in XRP balance, Account == root, issuer == root, currency == XRP
     std::map<key, STAmount> result;
 
     // populate a dictionary with low/high/currency/delta // this can be
```
The re-wrapping of the previous paragraph is a good thing.  But it's actually this following paragraph that has a couple of "`//`" in the middle of the paragraph that I suspect may be misplaced.

scottschurr commented on src/ripple/app/tests/Flow_test.cpp:
```diff
@@ -977,7 +977,7 @@ struct Flow_test : public beast::unit_test::suite
         auto const timeDelta = Env{*this}.closed ()->info ().closeTimeResolution;
 
-        for(auto const& t :{dcSoTime(), numAccSoTime()}){
+        for(auto const& t :{amendmentRIPD1141SoTime(), amendmentRIPD1141SoTime()}){
```
I'm probably missing something, but doesn't this outer `for` loop set `t` to the same value twice?  Do we still need the outer `for`?

seelabs replied:
You're right; That's not needed now that I just have one switchover over rather than two. Fixed.

vinniefalco replied:
`amendmentRIPD1141When`

vinniefalco commented:
FYI I'm not sure it was RIPD-1141, I just pulled that number out of a hat

scottschurr commented:
Looks good to me.  The most recent version builds on OS X for me and passes unit and npm tests.  :+1:

I'm not worried about a simple rename if RIPD-1141 ends up being the wrong JIRA entry.

seelabs commented:
@vinniefalco 1141 is one of the bugs that switch over fixes, and it's the most important one.

seelabs commented:
Rebased onto b18 and squashed, I'm setting this up to run overnight on my dogfood machine, but there were no substantial changes during the PR, so I don't anticipate any issues.

scottschurr commented:
sendMax path bug fix looks reasonable to me.  :+1:

scottschurr commented:
Nice to have the unit test.  :+1:  Do we know what AppVeyor is upset about?
